# Bài tập về nhà: Singular Value Decomposition (SVD)

Notebook này trình bày từng bước tính toán cho hai câu hỏi và dùng Python để kiểm tra các kết quả.

## Câu 1. Phân rã ma trận

Ta có

$$
A=
\begin{bmatrix}
1&0&1\\
0&1&1\\
1&0&1\\
0&1&1
\end{bmatrix}
=[\mathbf a_1\;\mathbf a_2\;\mathbf a_3].
$$

### 1. Sự phụ thuộc tuyến tính và rank của $A$

Ba cột thỏa mãn

$$\mathbf a_3=\mathbf a_1+\mathbf a_2,$$

hay $\mathbf a_1+\mathbf a_2-\mathbf a_3=\mathbf 0$. Vì vậy, các cột phụ thuộc tuyến tính. Hai cột $\mathbf a_1$ và $\mathbf a_2$ đều khác vector không và trực giao với nhau nên chúng độc lập tuyến tính. Do đó

$$\boxed{\operatorname{rank}(A)=2}.$$

### 2. Tính $A^\top A$

Tích vô hướng giữa các cột cho ta

$$
A^\top A=
\begin{bmatrix}
\mathbf a_1^\top\mathbf a_1&\mathbf a_1^\top\mathbf a_2&\mathbf a_1^\top\mathbf a_3\\
\mathbf a_2^\top\mathbf a_1&\mathbf a_2^\top\mathbf a_2&\mathbf a_2^\top\mathbf a_3\\
\mathbf a_3^\top\mathbf a_1&\mathbf a_3^\top\mathbf a_2&\mathbf a_3^\top\mathbf a_3
\end{bmatrix}
=
\begin{bmatrix}
2&0&2\\
0&2&2\\
2&2&4
\end{bmatrix}.
$$

In [6]:
import numpy as np
import pandas as pd
from IPython.display import display

A = np.array([[1, 0, 1], [0, 1, 1], [1, 0, 1], [0, 1, 1]], dtype=float)
AtA = A.T @ A
print("rank(A) =", np.linalg.matrix_rank(A))
print("A^T A =")
print(AtA)

rank(A) = 2
A^T A =
[[2. 0. 2.]
 [0. 2. 2.]
 [2. 2. 4.]]


### 3–6. Eigenvalues, eigenvectors và singular values

Đa thức đặc trưng của $A^\top A$ là

$$
\begin{aligned}
\det(A^\top A-\lambda I)
&=\det\begin{bmatrix}
2-\lambda&0&2\\
0&2-\lambda&2\\
2&2&4-\lambda
\end{bmatrix}\\
&=-\lambda(\lambda-2)(\lambda-6).
\end{aligned}
$$

Vì vậy, theo thứ tự giảm dần,

$$\lambda_1=6,\qquad \lambda_2=2,\qquad \lambda_3=0.$$

Giải $(A^\top A-\lambda_i I)\mathbf v_i=\mathbf 0$ rồi chuẩn hóa, ta chọn được các eigenvector trực chuẩn:

$$
\mathbf v_1=\frac{1}{\sqrt6}\begin{bmatrix}1\\1\\2\end{bmatrix},\qquad
\mathbf v_2=\frac{1}{\sqrt2}\begin{bmatrix}1\\-1\\0\end{bmatrix},\qquad
\mathbf v_3=\frac{1}{\sqrt3}\begin{bmatrix}-1\\-1\\1\end{bmatrix}.
$$

Thật vậy, chúng có độ dài bằng $1$, đôi một trực giao và thỏa mãn $(A^\top A)\mathbf v_i=\lambda_i\mathbf v_i$. Các singular value là

$$
\boxed{\sigma_1=\sqrt6,\qquad\sigma_2=\sqrt2,\qquad\sigma_3=0}.
$$

Các right singular vector chính là $\mathbf v_1,\mathbf v_2,\mathbf v_3$ ở trên. Dấu của một singular vector có thể đổi mà SVD vẫn đúng.

In [7]:
eigenvalues, eigenvectors = np.linalg.eigh(AtA)
order = np.argsort(eigenvalues)[::-1]
eigenvalues = eigenvalues[order]
eigenvectors = eigenvectors[:, order]
print("Eigenvalues giảm dần =", np.round(eigenvalues, 10))
print("Một bộ eigenvectors trực chuẩn theo cột =")
print(np.round(eigenvectors, 6))

Eigenvalues giảm dần = [6. 2. 0.]
Một bộ eigenvectors trực chuẩn theo cột =
[[-0.408248  0.707107  0.57735 ]
 [-0.408248 -0.707107  0.57735 ]
 [-0.816497  0.       -0.57735 ]]


### 7. Tính các left singular vector ứng với $\sigma_i>0$

Với $\sigma_1=\sqrt6$:

$$
\mathbf u_1=\frac{A\mathbf v_1}{\sigma_1}
=\frac{1}{2}\begin{bmatrix}1\\1\\1\\1\end{bmatrix}.
$$

Với $\sigma_2=\sqrt2$:

$$
\mathbf u_2=\frac{A\mathbf v_2}{\sigma_2}
=\frac{1}{2}\begin{bmatrix}1\\-1\\1\\-1\end{bmatrix}.
$$

### 8. Full SVD

Ta bổ sung hai vector đơn vị trực chuẩn, cùng trực giao với $\mathbf u_1$ và $\mathbf u_2$:

$$
\mathbf u_3=\frac{1}{\sqrt2}\begin{bmatrix}1\\0\\-1\\0\end{bmatrix},\qquad
\mathbf u_4=\frac{1}{\sqrt2}\begin{bmatrix}0\\1\\0\\-1\end{bmatrix}.
$$

Do đó, một Full SVD của $A$ là

$$
U=\begin{bmatrix}
\frac12&\frac12&\frac1{\sqrt2}&0\\
\frac12&-\frac12&0&\frac1{\sqrt2}\\
\frac12&\frac12&-\frac1{\sqrt2}&0\\
\frac12&-\frac12&0&-\frac1{\sqrt2}
\end{bmatrix},
\quad
\Sigma=\begin{bmatrix}
\sqrt6&0&0\\
0&\sqrt2&0\\
0&0&0\\
0&0&0
\end{bmatrix},
$$

$$
V^\top=\begin{bmatrix}
\frac1{\sqrt6}&\frac1{\sqrt6}&\frac2{\sqrt6}\\
\frac1{\sqrt2}&-\frac1{\sqrt2}&0\\
-\frac1{\sqrt3}&-\frac1{\sqrt3}&\frac1{\sqrt3}
\end{bmatrix}.
$$

Kích thước là $U\in\mathbb R^{4\times4}$, $\Sigma\in\mathbb R^{4\times3}$ và $V^\top\in\mathbb R^{3\times3}$.

### 9–10. Reduced SVD và kích thước

Vì $r=\operatorname{rank}(A)=2$, ta chỉ giữ hai singular value khác $0$ cùng các vector tương ứng:

$$
U_r=\begin{bmatrix}
\frac12&\frac12\\
\frac12&-\frac12\\
\frac12&\frac12\\
\frac12&-\frac12
\end{bmatrix},\qquad
\Sigma_r=\begin{bmatrix}\sqrt6&0\\0&\sqrt2\end{bmatrix},
$$

$$
V_r^\top=\begin{bmatrix}
\frac1{\sqrt6}&\frac1{\sqrt6}&\frac2{\sqrt6}\\
\frac1{\sqrt2}&-\frac1{\sqrt2}&0
\end{bmatrix}.
$$

Vậy

$$A=U_r\Sigma_rV_r^\top,$$

với $U_r\in\mathbb R^{4\times2}$, $\Sigma_r\in\mathbb R^{2\times2}$ và $V_r^\top\in\mathbb R^{2\times3}$.

In [8]:
sqrt = np.sqrt
U = np.array([[1/2, 1/2, 1/sqrt(2), 0], [1/2, -1/2, 0, 1/sqrt(2)], [1/2, 1/2, -1/sqrt(2), 0], [1/2, -1/2, 0, -1/sqrt(2)]])
Sigma = np.array([[sqrt(6), 0, 0], [0, sqrt(2), 0], [0, 0, 0], [0, 0, 0]])
V = np.array([[1/sqrt(6), 1/sqrt(2), -1/sqrt(3)], [1/sqrt(6), -1/sqrt(2), -1/sqrt(3)], [2/sqrt(6), 0, 1/sqrt(3)]])
U_r = U[:, :2]
Sigma_r = Sigma[:2, :2]
V_r_T = V[:, :2].T
print("U^T U =")
print(np.round(U.T @ U, 10))
print("V^T V =")
print(np.round(V.T @ V, 10))
print("U Sigma V^T =")
print(np.round(U @ Sigma @ V.T, 10))
print("U_r Sigma_r V_r^T =")
print(np.round(U_r @ Sigma_r @ V_r_T, 10))
print("Kích thước Reduced SVD =", U_r.shape, Sigma_r.shape, V_r_T.shape)

U^T U =
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
V^T V =
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
U Sigma V^T =
[[1. 0. 1.]
 [0. 1. 1.]
 [1. 0. 1.]
 [0. 1. 1.]]
U_r Sigma_r V_r^T =
[[1. 0. 1.]
 [0. 1. 1.]
 [1. 0. 1.]
 [0. 1. 1.]]
Kích thước Reduced SVD = (4, 2) (2, 2) (2, 3)


### 11. Kiểm tra

Kết quả Python phía trên cho thấy

$$U^\top U=I_4,\qquad V^\top V=I_3,$$

và cả Full SVD lẫn Reduced SVD đều tái tạo đúng ma trận $A$.

### 12. Giải thích

- Có một singular value bằng $0$ vì $A$ có ba cột nhưng chỉ có hai hướng độc lập, tức $\operatorname{rank}(A)=2$. Số singular value khác $0$ luôn bằng rank của ma trận.
- Right singular vector tương ứng là $\mathbf v_3$. Do $\sigma_3=0$, ta có $\|A\mathbf v_3\|^2=\mathbf v_3^\top A^\top A\mathbf v_3=0$, nên $A\mathbf v_3=\mathbf0$. Vì vậy, $\mathbf v_3$ là một cơ sở của null space $\mathcal N(A)$. Thật vậy, $A[-1,-1,1]^\top=\mathbf0$.
- Thành phần ứng với $\sigma_3=0$ chỉ tạo ra ma trận không: $\sigma_3\mathbf u_3\mathbf v_3^\top=0$. Reduced SVD bỏ thành phần này nên không làm mất thông tin và vẫn tái tạo chính xác $A$.

## Câu 2. Chọn $k$ cho Truncated SVD bằng tỷ lệ năng lượng

Các singular value là $12,5,3,1$.

### 1. Rank của $X$

Cả bốn singular value đều khác $0$, do đó

$$\boxed{\operatorname{rank}(X)=4}.$$

### 2. Tổng năng lượng

$$
E_{\text{total}}=12^2+5^2+3^2+1^2=144+25+9+1=\boxed{179}.
$$

### 3–4. Năng lượng giữ lại và tỷ lệ $R(k)$

$$
R(1)=\frac{144}{179}\approx80.45\%,
$$

$$
R(2)=\frac{144+25}{179}=\frac{169}{179}\approx94.41\%,
$$

$$
R(3)=\frac{144+25+9}{179}=\frac{178}{179}\approx99.44\%,
$$

$$
R(4)=\frac{179}{179}=100\%.
$$

In [9]:
singular_values = np.array([12, 5, 3, 1], dtype=float)
energies = singular_values ** 2
total_energy = energies.sum()
retained_energy = np.cumsum(energies)
ratios = retained_energy / total_energy
energy_table = pd.DataFrame({"k": [1, 2, 3, 4], "Năng lượng giữ lại": retained_energy.astype(int), "Tỷ lệ R(k)": ratios})
display(energy_table)

,k,Năng lượng giữ lại,Tỷ lệ R(k)
0,1,144,0.804469
1,2,169,0.944134
2,3,178,0.994413
3,4,179,1.000000


### 5. Chọn $k$ nhỏ nhất theo tỷ lệ năng lượng

Từ các tỷ lệ vừa tính:

| Mức năng lượng yêu cầu | $k$ nhỏ nhất | Tỷ lệ thật sự giữ lại |
|---:|---:|---:|
| 80% | 1 | 80.45% |
| 90% | 2 | 94.41% |
| 95% | 3 | 99.44% |
| 99% | 3 | 99.44% |

### 6–7. Sai số Frobenius

Với $k=1$:

$$\|X-X_1\|_F=\sqrt{5^2+3^2+1^2}=\sqrt{35}\approx5.9161.$$

Với $k=2$:

$$\|X-X_2\|_F=\sqrt{3^2+1^2}=\sqrt{10}\approx3.1623.$$

Với $k=3$:

$$\|X-X_3\|_F=\sqrt{1^2}=1.$$

In [10]:
errors = []
for k in [1, 2, 3]:
    error = np.sqrt(np.sum(singular_values[k:] ** 2))
    errors.append(error)
error_table = pd.DataFrame({"k": [1, 2, 3], "R(k)": ratios[:3], "||X - X_k||_F": errors})
display(error_table)

,k,R(k),||X - X_k||_F
0,1,0.804469,5.916080
1,2,0.944134,3.162278
2,3,0.994413,1.000000


### 8. Chọn $k$ khi $\sigma_4$ chủ yếu là nhiễu

Ta chọn $\boxed{k=3}$. Ba thành phần đầu đã giữ $178/179\approx99.44\%$ năng lượng, trong khi thành phần thứ tư chỉ chiếm $1/179\approx0.56\%$. Bỏ $\sigma_4$ giúp loại bớt nhiễu, giảm dữ liệu cần lưu và chỉ tạo sai số Frobenius bằng $1$.

### 9. Nhận xét

- **Vì sao singular values phải được sắp xếp giảm dần?** Cách sắp xếp này đặt hướng chứa nhiều năng lượng nhất lên trước. Vì thế, khi lấy $k$ thành phần đầu, ta nhận được xấp xỉ hạng $k$ tốt nhất theo chuẩn Frobenius và có thể tìm $k$ nhỏ nhất bằng tổng tích lũy.
- **Giữ càng nhiều singular values có phải lúc nào cũng tốt hơn không?** Nếu chỉ xét việc tái tạo đúng dữ liệu ban đầu thì giữ thêm thành phần sẽ không làm sai số tăng. Tuy nhiên, singular value nhỏ có thể chứa nhiễu; giữ chúng làm giảm khả năng nén, tốn bộ nhớ và đôi khi khiến mô hình học cả nhiễu. Vì vậy, giữ nhiều hơn không phải lúc nào cũng phù hợp với mục tiêu thực tế.
- **Sự đánh đổi giữa năng lượng và mức độ nén là gì?** Chọn $k$ lớn giữ nhiều năng lượng và giảm sai số nhưng nén ít hơn. Chọn $k$ nhỏ nén mạnh và tiết kiệm hơn nhưng bỏ nhiều thông tin hơn. Ta nên chọn $k$ nhỏ nhất vẫn đạt tỷ lệ năng lượng cần thiết và không giữ các thành phần chủ yếu là nhiễu.